sw/host/opentitanlib/src/transport/hyperdebug/dfu.rs

```diff
@@ -0,0 +1,407 @@
+                Ok(interface_name) => interface_name,
+                _ => continue,
+            };
+            if interface_desc.class_code() != USB_CLASS_APP
+                || interface_desc.sub_class_code() != USB_SUBCLASS_DFU
+                || (interface_desc.protocol_code() != 0x01
+                    && interface_desc.protocol_code() != 0x02)
+            {
+                continue;
+            }
+            dfu_interface = interface.number();
+            if let Some(extra_bytes) = interface_desc.extra() {
+                // Extra bytes contains inforation encoded according to DFU specification.
+                if extra_bytes.len() >= 9 {
+                    xfer_size = extra_bytes[5] as u32 | (extra_bytes[6] as u32) << 8;
+                }
+            }
+            lazy_static! {
+                static ref DFU_SECTION_REGEX: Regex =
+                    Regex::new("^@([^/]*)/0x([0-9a-fA-F]+)/([0-9]+)\\*([0-9]+)(..)").unwrap();
+            }
+            let Some(captures) = DFU_SECTION_REGEX.captures(&interface_name) else {
+                continue;
+            };
+            let section_name = captures.get(1).unwrap().as_str().trim();
+            if section_name != "Internal Flash" {
+                continue;
+            }
+            // We have the string describing the flash section (as opposed to fuses or
+            // once-programmable section), extract the relevant information.
+            base_address = u32::from_str_radix(captures.get(2).unwrap().as_str(), 16).unwrap();
+            let num_pages = captures.get(3).unwrap().as_str().parse::<u32>().unwrap();
+            page_size = captures.get(4).unwrap().as_str().parse::<u32>().unwrap();
+            let suffix = captures.get(5).unwrap().as_str();
+            if suffix.starts_with('K') {
+                page_size *= 1024;
+            }
+            flash_size = num_pages * page_size;
+        }
+    }
+    Ok(DfuDescriptor {
+        dfu_interface,
+        xfer_size,
+        page_size,
+        flash_size,
+        base_address,
+    })
+}
+
+/// Poll the bootloader using GETSTATUS request, until it leaves the "busy" state.
+fn wait_for_idle(dfu_device: &UsbBackend, dfu_interface: u8) -> Result<u8> {
+    loop {
+        let mut response = [0u8; 6];
+        let rc = dfu_device.read_control(
+            rusb::request_type(
+                rusb::Direction::In,
+                rusb::RequestType::Class,
+                rusb::Recipient::Interface,
+            ),
+            USB_DFU_GETSTATUS,
+            0,
+            dfu_interface as u16,
+            &mut response,
+        )?;
+        if rc != response.len() {
+            bail!(TransportError::FirmwareProgramFailed("".to_string()));
+        }
+        let command_status = response[0];
+        let minimum_delay_ms =
+            response[1] as u64 | (response[2] as u64) << 8 | (response[3] as u64) << 16;
```

Comment:
There is nothing wrong with constructing a u64 like this.  Depending on your personal preference, you could also write:
```
let foo = u64::from_le_bytes([response[1], response[2], response[3], 0, 0, 0, 0, 0]);
```

Reply:
That is a good suggestion.